Add guidelines and steps for creating a PR (#1060)
keptn/keptn.github.io: CONTRIBUTING.md

This documentation is authored with [markdown](https://www.markdownguide.org/basic-syntax/)
and rendered using the [Hugo static website generator framework](http://gohugo.io).

We welcome and encourage contributions of all levels.
You can make modifications using the github editor;
this works well for small modifications but, if you are making significant changes, you may find it better to fork and clone the repository and make changes using the text editor or IDE of your choice

> You can run the Hugo formatter locally to check the rendered documentation

Reviewer: ```suggestion
You can also run the Hugo based website locally to check the rendered documentation
```

Reviewer 2: You cannot make modifications using GitHub.com as the commits aren't signed and the current process complains if not all commits are signed.

and then push your changes to the repository as a PR.

> If you need help getting started, feel free to ask for help on the #keptn-help Slack channel.

Reviewer: ```suggestion
If you need help getting started, feel free to ask for help on the `#help-contributing` or `#keptn-docs` channels on [Keptn Slack?](https://keptn.sh/community/#slack).
```

We were all new to this once and are happy to help you!

## Guidelines for contributing

* Always fork the repository then clone that fork to your local system rather than clone master directly.
Keptn, like most open source projects, severely restricts who can push changes directly to the *master* branch to protect the integrity of the repository.
* Keep your language clean and crisp.
* Smaller PR's are easier to review and so generally get processed more quickly;
when possible, chunk your work into smallish PR's.
However, we recognize that documentation work sometimes requires larger PRs, such as when writing a whole new section or reorganizing existing files.

> * You may want to squash your commits before creating the final PR, to avoid conflicting commits.

Reviewer: ```suggestion
* You may want to squash your commits before creating the final PR, to avoid conflicting commits.
   It is **not mandatory**, the maintainers can always squash commits during the merge.
```

> * Be sure that the description of each commit and the PR itself are meaningful and described clearly.

Reviewer: Commits could be squash merged, so I do not see it as hard requirement

```suggestion
* Be sure that the description of the pull request itself are meaningful and described clearly.
```

This helps reviewers understand each commit and provides a good history after the PR is merged.

> * If your PR is not reviewed in a timely fashion, feel free to post a gentle reminder to the #keptn-docs Slack channel.

Reviewer: ```suggestion
* If your PR is not reviewed in a timely fashion, feel free to post a gentle reminder to the `#keptn-docs` Slack channel.
```

* Resolve review comments and suggestions promptly.

> If you see a problem and are unable to fix it yourself or have an idea for an enhancement, please create an issue on the github repository:

Reviewer: ```suggestion
If you see a problem and are unable to fix it yourself or have an idea for an enhancement, please create an issue on the GitHub repository:
```

* Provide specific and detailed information about the problem and possible solutions to help others understand the issue.
* When reporting a bug, provide a detailed list of steps to reach the bug.
If possible, also attach screenshots that illustrate the bug.

> * If you want to do the work on an issue, include that information in your description of the issue.

Reviewer: NIT: Ideally should be handled by issues templates. There is a support for forms introduced recently, can show it to you next week. E.g. see https://github.com/jenkins-infra/helpdesk/tree/main/.github/ISSUE_TEMPLATE by @lemeurherve

Author: Cool!  After you show me how to use the issues template, I'll augment this doc to include instructions.

This documentation is authored with [markdown](https://www.markdownguide.org/basic-syntax/)
and rendered using the [Hugo static website generator framework](http://gohugo.io).

We welcome and encourage contributions of all levels.

> You can make modifications using the github editor;

Reviewer 3: The github editor doesn't work that well for keptn because we have the requirement for the DCO which is not available in the GH editor AFAIK

Reviewer 3: I don't know if this can be fixed by replacing the DCO. It really depends on the replacement IMO.

Reviewer 3: The DCO needs to be attached to every single commit you make, so there's no nice way to get around it. You can however add the DCO to the commit messages after the fact of course by doing a rebase of your branch.

Reviewer 4: The DCO states: `By making a contribution to this project, I certify that: ....`

Keptn is therefore compliant by ensuring that "by making a contribution via PR to this project, you are implicitely agreeing to the DCO". Once and done?

Reviewer 3: I fully agree that the DCO has no real purpose whatsoever since it's just some random text that is added somewhere. The problem is rather that for now, we have the requirement by the CNCF to have it...
So my suggestion would be to only document workflows that actually work for us right now.

Author: I think Oleg has something in mind that will fix this -- and other -- problems.  I added a mealy-mouth warning that using the github editor causes the CI to fail and I think that is good enough for now.  See if you agree.

Reviewer: Yes, I am working on it. FTR https://github.com/keptn/community/issues/72

this works well for small modifications but, if you are making significant changes, you may find it better to fork and clone the repository and make changes using the text editor or IDE of your choice

> You can also run the Hugo based website locally to check the rendered documentation

Reviewer 3: ```suggestion
You can also run the [Hugo](https://gohugo.io/) based website locally to check the rendered documentation
```

> and then push your changes to the repository as a PR.

Reviewer 3: ```suggestion
and then push your changes to the repository as a pull request.
```

> If you need help getting started, feel free to ask for help on the `#help-contributing` or `#keptn-docs` channels on [Keptn Slack?](https://keptn.sh/community/#slack).

Reviewer 3: ```suggestion
If you need help getting started, feel free to ask for help on the `#help-contributing` or `#keptn-docs` channels on the [Keptn Slack](https://keptn.sh/community/#slack).
```

We were all new to this once and are happy to help you!

## Guidelines for contributing

* Always fork the repository then clone that fork to your local system rather than clone master directly.
Keptn, like most open source projects, severely restricts who can push changes directly to the *master* branch to protect the integrity of the repository.
* Keep your language clean and crisp.
* Smaller PR's are easier to review and so generally get processed more quickly;
when possible, chunk your work into smallish PR's.
However, we recognize that documentation work sometimes requires larger PRs, such as when writing a whole new section or reorganizing existing files.
* You may want to squash your commits before creating the final PR, to avoid conflicting commits.

>    It is **not mandatory**, the maintainers can always squash commits during the merge.

Reviewer 3: ```suggestion
   It is **not mandatory**, the maintainers will squash your commits during the merge anyways.
```

Reviewer 3: we always squash & merge in keptn :)

> * Be sure that the description of the pull request itself are meaningful and described clearly.

Reviewer 3: ```suggestion
* Be sure that the description of the pull request itself is meaningful and described clearly.
```

This helps reviewers understand each commit and provides a good history after the PR is merged.
* If your PR is not reviewed in a timely fashion, feel free to post a gentle reminder to the `#keptn-docs` Slack channel.
* Resolve review comments and suggestions promptly.

If you see a problem and are unable to fix it yourself or have an idea for an enhancement, please create an issue on the GitHub repository:

* Provide specific and detailed information about the problem and possible solutions to help others understand the issue.

> * When reporting a bug, provide a detailed list of steps to reach the bug.

Reviewer 3: ```suggestion
* When reporting a bug, provide a detailed list of steps to reproduce the bug.
```

* If you want to do the work on an issue, include that information in your description of the issue.

> =======

Author: ```suggestion
```

   :warning: It is important that you install the **extended** version of Hugo. Learn how to install Hugo, depending on your OS, here: [install Hugo](https://gohugo.io/getting-started/installing/).

The themes directory on your local machine (localdirectory/keptn.github.io/themes) is empty.

> The Hugo serif theme is available at https://github.com/zerostaticthemes/hugo-serif-theme.

Reviewer: I would actually move it to the top of the guide where Hugo is referenced

```suggestion
Currently the website uses the _Hugo serif theme_ is available at https://github.com/zerostaticthemes/hugo-serif-theme.
```

You need to load the git submodule (see next step) to install this theme.

> 2. Load the git submodule for the Hugo serif theme in the *themes* directory:

Reviewer: Just in case we change it later

```suggestion
2. Load the git submodule for the Hugo theme in the *themes* directory:
```

    ```
    git submodule update --init --recursive --force
    ```
3. Execute the following command from the root folder of your clone:
    ```
    hugo server -D
4. Start contributing!
Leave the Hugo server running in a shell.

> Note that all edits to the files are updated immediately.

Author: ```suggestion
Note that Hugo updates the rendered documentation each time you write the file.
```

   :warning: It is important that you install the **extended** version of Hugo. Learn how to install Hugo, depending on your OS, here: [install Hugo](https://gohugo.io/getting-started/installing/).

The themes directory on your local machine (localdirectory/keptn.github.io/themes) is empty.

> Currently the website uses the _Hugo serif theme_ is available at https://github.com/zerostaticthemes/hugo-serif-theme.

Reviewer 3: ```suggestion
Currently the website uses the _Hugo serif theme_ which is available at https://github.com/zerostaticthemes/hugo-serif-theme.
```

You need to load the git submodule (see next step) to install this theme.

2. Load the git submodule for the Hugo theme in the *themes* directory:
    ```
    git submodule update --init --recursive --force
    ```
3. Execute the following command from the root folder of your clone:
    ```

>     hugo server -D

Reviewer 3: ```suggestion
    hugo server -D
    ```
```

4. Start contributing!
Leave the Hugo server running in a shell.
Note that Hugo updates the rendered documentation each time you write the file.

>     ```

Reviewer 3: ```suggestion
    
```

Each topical subdirectory contains:

  - An *index.md* file that has the metadata discussed above plus the text for the section
  - An *assets* subdirectory where graphical files for that topic are stored.
No *assets* subdirectory is present if the topic has no graphics.

> ## Submitting new content through a PR

Reviewer: ```suggestion
## Submitting new content through a pull request
```

If you have forked and cloned the repository,
you can modify the documentation or add new material
by editing the markdown file(s) in the local clone of your fork

> and then submitting a PR.

Reviewer: ```suggestion
and then submitting a _pull request (PR)_.
```

Note that you can also modify the source using the github editor.
This is very useful when you want to fix a typo or make a small editorial change but, if you are doing significant writing, it is generally better to work on files in your local clone.

The following sequence of steps is a reasonable workflow for creating new content on your local clone and pushing it to github to be reviewed and merged.

1. Execute the following command from the root folder and leave it running in a shell:
    ```
    hugo server -D
    ```
2. Create a local branch for your changes.  Be sure to base your new branch on the contents of the `master` branch unless you intend to create a derivative PR:
   ```

>    git branch master

Reviewer: Some of these bits could be simplified via GitHub CLI: https://cli.github.com/
Maybe it is something to recommend for newcomer contributors in the next iteration of the guidelines

  - An *index.md* file that has the metadata discussed above plus the text for the section
  - An *assets* subdirectory where graphical files for that topic are stored.
No *assets* subdirectory is present if the topic has no graphics.

## Submitting new content through a pull request

If you have forked and cloned the repository,
you can modify the documentation or add new material
by editing the markdown file(s) in the local clone of your fork
and then submitting a _pull request (PR)_.

Note that you can also modify the source using the github editor.
This is very useful when you want to fix a typo or make a small editorial change but, if you are doing significant writing, it is generally better to work on files in your local clone.

The following sequence of steps is a reasonable workflow for creating new content on your local clone and pushing it to github to be reviewed and merged.

1. Execute the following command from the root folder and leave it running in a shell:
    ```
    hugo server -D
    ```
2. Create a local branch for your changes.  Be sure to base your new branch on the contents of the `master` branch unless you intend to create a derivative PR:
   ```

>    git branch master

Reviewer 3: ```suggestion
   git checkout master
```

   git pull
   git checkout -b <your-new-branch>
   ```
3. Execute the following and check the output to ensure that your branch is set up correctly:
   ```
   git status
   ```

> 3. Do the writing you want to do in your local branch, checking the formatted version at localhost:1313.

Reviewer 3: ```suggestion
3. Do the writing you want to do in your local branch, checking the formatted version at `localhost:1313`.
```

4. When you have completed the writing you want to do, close all files in your branch and run `git status` to confirm that it correctly reflects the files you have modified, added, and deleted.

4. Add and commit your changes.  Here, we commit all modified files but you can specify individual files to the `git add` command.
The `git commit -s` command commits the files and signs that you are contributing this intellectual property to the Keptn project.
   ```
   git add .
   git commit -s
   ```

>    Use vi commands to add a description of the PR (should be 80 characters or less) to the commit.  This will be displayed as the title of the PR in listings.

Reviewer 3: ```suggestion
   Use vi commands to add a description of the PR (should be 80 characters or less) to the commit. This will be displayed as the title of the PR in listings.
```

You can add multiple lines explaining the PR here but, in general, it is better to only supply the PR title here; you can add more information and edit the PR title when you create the PR on the github UI page.

5. Push your branch to github:
   - If you cloned your fork to use SSH, the command is:
     ```
     git push --set-upstream origin <branch-name>
     ```

     Note that you can just issue the `git push` command.
Git responds with an error message that gives you the full command line to use; you can copy this command and paste it into your shell to push the content.
   - If you cloned your fork to use HTTP, the command is:
     ```
     git push <need options/arguments>
     ```

> 6. Create the PR by going to the [keptn.github.io](https://github.com/keptn/keptn.github.io) web site.

Reviewer 3: ```suggestion
6. Create the PR by going to the [keptn.github.io](https://github.com/keptn/keptn.github.io) GitHub repository.
```

     ```
     <GitID>:<branch> had recent pushes less than a minute ago
     ```

   - Click on the button in that shaded area marked:
     ```
     Compare & pull request
     ```
   - Check that the title of the PR is correct; click the "Edit" button to modify it.
Add "WIP" (Work in Progress) or "Draft" to the title if the PR is not yet ready for general review.
   - Add a full description of the work in the PR, including any notes for reviewers, a reference to the relevant github issue (if any), and tags for specific people (if any) who may be interested in this PR.
   - Carefully review the changes github displays for this PR to ensure that they are what you want.
   - Click the green "Create pull request" button to create the PR.
You may want to record the PR number somewhere for future reference although you can always find the PR in the github lists of open and closed PRs.
   - Github automatically populates the "Reviewers" block.
   - If this PR is not ready for review, click the "Still in progress? Convert to draft" string under the list of reviewers.
People can still review the content but can not merge the PR until you remove the "Draft" status.

>    - The block of the PR that reports on checks and such will include the following item:

Reviewer 3: ```suggestion
   - The block of the PR that reports on checks will include the following item:
```

     ```
     This pull request is still a work in progress
     Draft pull requests cannot be merged.
     ```

>    - When the PR is ready to be reviewed, approved, and merged, click the "Ready to review" box next to remove the "Draft" status.

Reviewer 3: ```suggestion
   - When the PR is ready to be reviewed, approved, and merged, click the "Ready to review" button to remove the "Draft" status. Then, if you added "WIP" or "Draft" to the PR title, remove it now.
```
